src/community/wfs3/src/main/java/org/geoserver/wfs3/response/RFCGeoJSONFeaturesResponse.java

```diff
@@ -0,0 +1,189 @@
+/* (c) 2018 Open Source Geospatial Foundation - all rights reserved
+ * This code is licensed under the GPL 2.0 license, available at the root
+ * application directory.
+ */
+package org.geoserver.wfs3.response;
+
+import org.geoserver.catalog.FeatureTypeInfo;
+import org.geoserver.config.GeoServer;
+import org.geoserver.ows.Dispatcher;
+import org.geoserver.ows.Request;
+import org.geoserver.ows.URLMangler;
+import org.geoserver.ows.util.ResponseUtils;
+import org.geoserver.platform.Operation;
+import org.geoserver.platform.ServiceException;
+import org.geoserver.wfs.json.GeoJSONBuilder;
+import org.geoserver.wfs.json.GeoJSONGetFeatureResponse;
+import org.geoserver.wfs.request.FeatureCollectionResponse;
+import org.geoserver.wfs.request.GetFeatureRequest;
+import org.geoserver.wfs.request.Query;
+import org.geoserver.wfs3.NCNameResourceCodec;
+import org.geotools.feature.FeatureCollection;
+import org.geotools.feature.NameImpl;
+import org.geotools.referencing.CRS;
+import org.geotools.referencing.crs.DefaultGeographicCRS;
+import org.geotools.util.Version;
+import org.opengis.feature.Feature;
+import org.opengis.referencing.crs.CoordinateReferenceSystem;
+
+import javax.xml.namespace.QName;
+import java.io.BufferedWriter;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.OutputStreamWriter;
+import java.io.UnsupportedEncodingException;
+import java.math.BigInteger;
+import java.util.Collections;
+import java.util.List;
+
+import static org.geoserver.wfs3.DefaultWebFeatureService30.getAvailableFormats;
+
+public class RFCGeoJSONFeaturesResponse extends GeoJSONGetFeatureResponse {
+
+    static final ThreadLocal<String> WFS3_FEATURE_ID = new ThreadLocal<String>();
+
+    /** The MIME type requested by WFS3 for GeoJSON Responses */
+    public static final String MIME = "application/geo+json";
+
+    public RFCGeoJSONFeaturesResponse(GeoServer gs) {
+        super(gs, MIME);
+    }
+
+    @Override
+    public String getMimeType(Object value, Operation operation) throws ServiceException {
+        return MIME;
+    }
+
+    public void write(Object value, OutputStream output, Operation operation) throws IOException {
+        // was it a single feature request?
+        String requestFeatureId = getWFS3FeatureId();
+        if (requestFeatureId != null) {
```

Comment:
I guess the difference is that when there is only a single feature no array is used ? I dislike this, it will make client life more difficult for no clear reason, but well we need to respect the standard.

Reply:
Go and complain with the spec writers, when there is a single feature, that's what the spec orders to to.